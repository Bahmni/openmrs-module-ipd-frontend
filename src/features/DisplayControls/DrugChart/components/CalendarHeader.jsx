import React from "react";
import PropTypes from "prop-types";
import "../styles/CalendarHeader.scss";
import data from "../../../../utils/config.json";

export default function CalendarHeader(props) {
  const { currentShiftArray } = props;
  const enable24hour = data.config.drugChart.enable24HourTime;

  return (
    <div className="calendar-header">
      <div style={{ display: "flex" }}>
        {currentShiftArray.map((hour) => {
          const transformedHour = enable24hour ? hour : hour % 12 || 12;
          const formattedHour = `${transformedHour.toLocaleString("en-US", {
            minimumIntegerDigits: 2,
          })}:00`;

          return (
            <div data-testid="hour" key={hour} className={"hour-header"}>
              {formattedHour}
            </div>
          );
        })}
      </div>
    </div>
  );
}
CalendarHeader.propTypes = {
  currentShiftArray: PropTypes.array,
};
