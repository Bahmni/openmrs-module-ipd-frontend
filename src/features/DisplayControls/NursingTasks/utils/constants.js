export const items = [
  { id: "allTasks", text: "All Tasks" },
  { id: "completed", text: "Completed" },
  { id: "pending", text: "Pending" },
  { id: "stopped", text: "Stopped" },
];
