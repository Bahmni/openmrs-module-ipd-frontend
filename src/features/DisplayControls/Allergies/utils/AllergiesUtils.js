import axios from "axios";
import { ALLERGIES_BASE_URL } from "../../../../constants";

export const fetchAllergiesData = async (patientUuid) => {
  const FETCH_ALLERGIES_URL = `${ALLERGIES_BASE_URL}?patient=${patientUuid}&_summary=data`;
  try {
    const response = await axios.get(FETCH_ALLERGIES_URL);
    return response;
  } catch (error) {
    return error;
  }
};
