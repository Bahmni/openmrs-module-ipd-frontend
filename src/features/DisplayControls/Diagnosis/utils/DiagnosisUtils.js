import axios from "axios";
import { DIAGNOSIS_SEARCH_URL } from "../../../../constants";

export const getPatientDiagnosis = async (patientUuid) => {
  try {
    const response = await axios.get(DIAGNOSIS_SEARCH_URL, {
      params: { patientUuid: patientUuid },
      withCredentials: true,
    });
    return response.data;
  } catch (error) {
    return error;
  }
};

export const diagnosisHeaders = [
  {
    id: "1",
    header: "Diagnosis",
    key: "diagnosis",
    isSortable: false,
  },
  {
    id: "2",
    header: "Order",
    key: "order",
    isSortable: true,
  },
  {
    id: "3",
    header: "Certainty",
    key: "certainty",
    isSortable: true,
  },
  {
    id: "4",
    header: "Status",
    key: "status",
    isSortable: true,
  },
  {
    id: "5",
    header: "Diagnosed By",
    key: "diagnosedBy",
    isSortable: false,
  },
  {
    id: "6",
    header: "Diagnosis Date ",
    key: "diagnosisDate",
    isSortable: true,
  },
];
