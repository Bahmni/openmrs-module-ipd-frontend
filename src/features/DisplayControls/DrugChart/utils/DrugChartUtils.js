import axios from "axios";
import moment from "moment";
import { MEDICATIONS_BASE_URL } from "../../../../constants";
import data from "../../../../utils/config.json";
import { AdminMedicationData } from "../components/AdminMedicationData";

const { config: { drugChart = {} } = {} } = data;

export const fetchMedications = async (patientUuid, forDate) => {
  const FETCH_MEDICATIONS_URL = `${MEDICATIONS_BASE_URL}?patientUuid=${patientUuid}&forDate=${forDate}`;
  try {
    const response = await axios.get(FETCH_MEDICATIONS_URL);
    return response;
  } catch (error) {
    console.error(error);
  }
};

const isLateTask = (startTime) => {
  const currentTime = Math.floor(new Date().getTime() / 1000);
  const lateTaskStatusWindowInSeconds =
    drugChart.timeInMinutesFromNowToShowPastTaskAsLate * 60;

  return startTime < currentTime - lateTaskStatusWindowInSeconds;
};

const isAdministeredLateTask = (startTime, effectiveStartDate) => {
  const lateTaskStatusWindowInSeconds =
    drugChart.timeInMinutesFromStartTimeToShowAdministeredTaskAsLate * 60;

  return effectiveStartDate - startTime > lateTaskStatusWindowInSeconds;
};

const checkIfSlotIsAdministered = (status) => {
  return status === "COMPLETED";
};

export const getUTCEpochForDate = (viewDate) => {
  const utcTimeEpoch = moment.utc(viewDate).unix();
  return utcTimeEpoch;
};

export const TransformDrugChartData = (drugChartData) => {
  const drugOrderData = [];
  const slotDataByOrder = [];
  console.log("drugChartData", drugChartData);

  AdminMedicationData.map((schedule) => {
    const { slots } = schedule;

    const administeredTimeInfo = [];

    slots.forEach((slot) => {
      const slotData = {};
      const { startTime, status, order, medicationAdministration } = slot;
      let medicationStatus = "Pending";
      let adminInfo = "",
        administeredTime;

      const isCompleted = checkIfSlotIsAdministered(status);

      if (isCompleted) {
        const isLate = isAdministeredLateTask(
          startTime,
          medicationAdministration.administeredDateTime
        );
        medicationStatus = isLate ? "Administered-Late" : "Administered";
        if (medicationAdministration) {
          const { administeredDateTime, provider } = medicationAdministration;
          const administeredDateTimeObject = new Date(
            administeredDateTime * 1000
          );
          administeredTime = moment(administeredDateTimeObject).format("HH:mm");
          adminInfo = provider.display + " [" + administeredTime + "]";
        } else {
          adminInfo = "";
        }
      }

      const drugOrder = {
        uuid: order.uuid,
        drugName: order.drug.display,
        drugRoute: order.route.display,
        administrationInfo: administeredTimeInfo,
        dosingInstructions: order.dosingInstructions,
      };

      if (order.duration) {
        drugOrder.duration = order.duration + " " + order.durationUnits.display;
      }
      if (order.doseUnits.display !== "ml") {
        drugOrder.dosage = order.dose;
        drugOrder.doseType = order.doseUnits.display;
      } else {
        drugOrder.dosage = order.dose + order.doseUnits.display;
      }
      if (order.duration) {
        drugOrder.duration = order.duration + " " + order.durationUnits.display;
      }

      const startDateTimeObj = new Date(startTime * 1000);

      const setLateStatus = isLateTask(startTime);
      const startHour = startDateTimeObj.getHours();
      const startMinutes = startDateTimeObj.getMinutes();
      slotData[startHour] = {
        minutes: startMinutes,
        status: !isCompleted && setLateStatus ? "Late" : medicationStatus,
        administrationInfo: adminInfo,
      };
      if (
        medicationStatus === "Administered" ||
        medicationStatus === "Administered-Late"
      ) {
        const adminData = {
          kind: medicationStatus,
          time: administeredTime,
        };
        drugOrder.administrationInfo.push(adminData);
      }
      if (
        !drugOrderData.some(
          (existingOrder) =>
            existingOrder.drugName === drugOrder.drugName &&
            existingOrder.uuid === drugOrder.uuid
        )
      ) {
        drugOrderData.push(drugOrder);
        slotDataByOrder.push(slotData);
      } else {
        const index = drugOrderData.findIndex(
          (existingOrder) =>
            existingOrder.drugName === drugOrder.drugName &&
            existingOrder.uuid === drugOrder.uuid
        );
        slotDataByOrder[index] = {
          ...slotDataByOrder[index],
          ...slotData,
        };
      }
    });
  });
  return [slotDataByOrder, drugOrderData];
};
