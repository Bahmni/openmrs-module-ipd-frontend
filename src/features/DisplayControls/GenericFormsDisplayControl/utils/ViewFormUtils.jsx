import axios from "axios";
import { FETCH_ALL_OBSERVATIONS_IN_ENCOUNTER_URL } from "../../../../constants";

export const fetchAllObservations = async (encounterUuid) => {
  try {
    const response = await axios.get(
      FETCH_ALL_OBSERVATIONS_IN_ENCOUNTER_URL.replace(
        "{encounterUuid}",
        encounterUuid
      )
    );
    if (response.status === 200) {
      return response.data;
    }
    return [];
  } catch (error) {
    console.error(error);
  }
};
export const flattenObservations = (observations) => {
  let flatObservations = [];
  observations.forEach((observation) => {
    flatObservations.push(observation);
    if (observation.groupMembers && observation.groupMembers.length > 0) {
      const flatGroupMembers = flattenObservations(observation.groupMembers);
      flatObservations = flatObservations.concat(flatGroupMembers);
    }
  });
  return flatObservations;
};
