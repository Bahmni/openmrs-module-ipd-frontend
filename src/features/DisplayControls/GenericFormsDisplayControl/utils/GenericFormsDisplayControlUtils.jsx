import axios from "axios";
import { FETCH_FORM_DETAILS_URL } from "../../../../constants";
import { getFetchFormTranslationsUrl } from "../../../../utils/CommonUtils";

export const fetchAllConceptsForForm = async (formUuid) => {
  try {
    const response = await axios.get(
      FETCH_FORM_DETAILS_URL.replace("{formUuid}", formUuid),
      {
        withCredentials: true,
      }
    );
    return response.data;
  } catch (e) {
    console.error(e);
  }
};

export const fetchFormTranslations = async (formName, formUuid) => {
  try {
    const response = await axios.get(
      getFetchFormTranslationsUrl(formName.replace(" ", "+"), formUuid),
      {
        withCredentials: true,
      }
    );
    return response.data;
  } catch (e) {
    console.error(e);
  }
};
