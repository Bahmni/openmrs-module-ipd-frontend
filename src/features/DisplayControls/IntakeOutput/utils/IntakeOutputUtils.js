import React from "react";
import { TooltipCarbon } from "bahmni-carbon-ui";
import NoteIcon from "../../../../icons/note.svg";
import {
  defaultDateTimeFormat,
  BAHMNI_CORE_OBSERVATIONS_BASE_URL,
} from "../../../../constants";
import { formatDate } from "../../../../utils/DateTimeUtils";
import { FormattedMessage } from "react-intl";
import moment from "moment";
import axios from "axios";

export const filterDataForRange = (
  data,
  timeConceptNames,
  startEndDates,
  visitStartDateTime
) => {
  return data.filter((obs) => {
    if (obs.visitStartDateTime === visitStartDateTime) {
      const timeConcept = obs.groupMembers.find((member) =>
        timeConceptNames?.includes(member?.concept?.name)
      );
      const obsTime = moment(timeConcept?.value).startOf("minute");
      const startTime = moment(startEndDates.startDate).startOf("minute");
      const endTime = moment(startEndDates.endDate).startOf("minute");
      return obsTime.isBetween(startTime, endTime, null, "[)");
    }
  });
};

export const getSortedObsData = (obsData, sortConceptNames) => {
  const sortedData = [...obsData].sort((a, b) => {
    const sortDataA = a.groupMembers.find((member) =>
      sortConceptNames?.includes(member?.concept?.name)
    );
    const sortDataB = b.groupMembers.find((member) =>
      sortConceptNames?.includes(member?.concept?.name)
    );

    if (sortDataA && sortDataB) {
      return (
        new Date(sortDataA.valueAsString) - new Date(sortDataB.valueAsString)
      );
    } else if (sortDataA) {
      return -1;
    } else if (sortDataB) {
      return 1;
    } else {
      return 0;
    }
  });
  return sortedData;
};

export const transformObsData = (obsData, config) => {
  const consolidatedIOData = {};
  const transformedData = [];
  let balance = 0.0;
  let quantityUnits = "";
  const totalIntakeMap = {};
  const totalOutputMap = {};
  obsData.forEach((obs) => {
    const dateTime = obs.groupMembers.find((member) =>
      config?.timeConceptNames?.includes(member?.concept?.name)
    );
    const intakeRoute = obs.groupMembers.find(
      (member) => member?.concept?.name === config.intakeRouteConcept
    );
    const intakeQuantity = obs.groupMembers.find(
      (member) => member?.concept?.name === config.intakeQuantityConcept
    );
    const outputRoute = obs.groupMembers.find(
      (member) => member?.concept?.name === config.outputRouteConcept
    );
    const outputQuantity = obs.groupMembers.find(
      (member) => member?.concept?.name === config.outputQuantityConcept
    );
    quantityUnits = intakeQuantity
      ? intakeQuantity?.concept?.units
      : outputQuantity?.concept?.units;
    const getBalance = () => {
      balance = caculateBalance(
        intakeQuantity?.value,
        outputQuantity?.value,
        balance
      );
      if (
        balance > config?.normalHighValue ||
        balance < config?.normalLowValue
      ) {
        return (
          <span className={"red-text"}>{balance + " " + quantityUnits}</span>
        );
      } else {
        return <span>{balance + " " + quantityUnits}</span>;
      }
    };
    transformedData.push({
      id: obs.uuid,
      dateTime: formatDate(dateTime?.value, defaultDateTimeFormat),
      intakeRoute: intakeRoute && getRoute(intakeRoute),
      intakeQuantity:
        intakeQuantity && getQuantity(intakeQuantity, quantityUnits),
      outputRoute: outputRoute && getRoute(outputRoute),
      outputQuantity:
        outputQuantity && getQuantity(outputQuantity, quantityUnits),
      balance: getBalance(),
    });
    let intakeGroup = totalIntakeMap[intakeRoute?.valueAsString];
    if (intakeRoute && intakeGroup) {
      intakeQuantity &&
        (intakeGroup.quantity += intakeQuantity && intakeQuantity?.value);
      totalIntakeMap[intakeRoute?.valueAsString] = intakeGroup;
    } else if (intakeRoute) {
      intakeGroup = {};
      intakeGroup.quantity = intakeQuantity && intakeQuantity?.value;
      intakeGroup.quantityUnits = quantityUnits;
      totalIntakeMap[intakeRoute.valueAsString] = intakeGroup;
    }
    let outputGroup = totalOutputMap[outputRoute?.valueAsString];
    if (outputRoute && outputGroup) {
      outputQuantity &&
        (outputGroup.quantity += outputQuantity && outputQuantity?.value);
      totalOutputMap[outputRoute?.valueAsString] = outputGroup;
    } else if (outputRoute) {
      outputGroup = {};
      outputGroup.quantity = outputQuantity && outputQuantity?.value;
      outputGroup.quantityUnits = quantityUnits;
      totalOutputMap[outputRoute.valueAsString] = outputGroup;
    }
  });
  consolidatedIOData.transformedData = transformedData;
  consolidatedIOData.totalIntakeMap = totalIntakeMap;
  consolidatedIOData.totalOutputMap = totalOutputMap;
  return consolidatedIOData;
};

export const caculateBalance = (input, output, balance) => {
  input && (balance += input);
  output && (balance -= output);
  return balance;
};

export const getRoute = (route) => {
  return (
    <div className="notes-icon-div">
      {route?.comment && route?.comment !== "" && (
        <TooltipCarbon
          className="notes-icon"
          icon={() => <NoteIcon />}
          content={route?.comment}
        />
      )}
      <span className={"concept-name"}>{route?.valueAsString}</span>
    </div>
  );
};

export const getQuantity = (quantity, quantityUnits) => {
  return (
    <div className="notes-icon-div">
      {quantity?.comment && quantity?.comment !== "" && (
        <TooltipCarbon
          className="notes-icon"
          icon={() => <NoteIcon />}
          content={quantity?.comment}
        />
      )}
      <span className={"concept-name"}>
        {quantity?.value?.toString() + " " + quantityUnits}
      </span>
    </div>
  );
};

export const intakeOutputHeaders = [
  {
    header: "Date and Time",
    key: "dateTime",
  },
  {
    header: "Intake Route",
    key: "intakeRoute",
  },
  {
    header: "Quantity",
    key: "intakeQuantity",
  },
  {
    header: "Output Route",
    key: "outputRoute",
  },
  {
    header: "Quantity",
    key: "outputQuantity",
  },
  {
    header: "Balance",
    key: "balance",
  },
];

export const currentPeriodRange = (date, periodConfig = {}) => {
  const currentDate = moment(date);
  const [hours, minutes] = periodConfig.startTime.split(":");
  const startDate = moment(date).clone().hour(hours).minute(minutes).second(0);
  const endDate = moment(startDate)
    .clone()
    .add(periodConfig.durationInHours, "hours");
  const currentRange = {};
  if (currentDate.isBetween(startDate, endDate, null, "[]")) {
    currentRange.startDateTime = startDate;
    currentRange.endDateTime = endDate;
  } else if (currentDate.isSameOrBefore(startDate)) {
    currentRange.startDateTime = moment(startDate)
      .clone()
      .subtract(periodConfig.durationInHours, "hours");
    currentRange.endDateTime = moment(currentRange.startDateTime)
      .clone()
      .add(periodConfig.durationInHours, "hours");
  }
  return currentRange;
};

export const NotCurrentPeriodMessage = (
  <FormattedMessage
    id={"NOT_CURRENT_PERIOD_MESSAGE"}
    defaultMessage={"You're not viewing the current period"}
  />
);

export const NoDataForSelectedPeriod = (
  <FormattedMessage
    id={"NO_IO_DATA_MESSAGE"}
    defaultMessage={
      "No intake output data is available for the patient in this period"
    }
  />
);

export const isCurrentPeriod = (startEndDates, periodConfig = {}) => {
  const currentPeriod = currentPeriodRange(new Date(), periodConfig);
  return (
    moment(currentPeriod.startDateTime)
      .startOf("minute")
      .isSame(moment(startEndDates.startDate).startOf("minute")) &&
    moment(currentPeriod.endDateTime)
      .startOf("minute")
      .isSame(moment(startEndDates.endDate).startOf("minute"))
  );
};

export const fetchFormData = async (
  patientUuid,
  conceptNames,
  numberOfVisits
) => {
  const conceptParams = conceptNames
    .map((concept) => `concept=${concept.replace(" ", "+")}`)
    .join("&");
  const INTAKE_OUTPUT_URL = `${BAHMNI_CORE_OBSERVATIONS_BASE_URL}${conceptParams}&numberOfVisits=${numberOfVisits}&patientUuid=${patientUuid}`;
  try {
    const response = await axios.get(INTAKE_OUTPUT_URL, {
      withCredentials: true,
    });
    if (response.status !== 200) throw new Error(response.statusText);
    return response.data;
  } catch (error) {
    return error;
  }
};
