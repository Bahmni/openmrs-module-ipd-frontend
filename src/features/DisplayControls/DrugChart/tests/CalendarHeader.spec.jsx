import React from "react";
import { render } from "@testing-library/react";
import CalendarHeader from "../components/CalendarHeader";
import { currentShiftHoursArray } from "../utils/DrugChartUtils";
import MockDate from "mockdate";

describe("CalendarHeader", () => {
  it("should match snapshot", () => {
    MockDate.set("2023-12-22T00:00:00.000+0530");
    const currentShiftArray = [14, 15, 16, 17, 18, 19, 20, 21];
    const { container } = render(
      <CalendarHeader currentShiftArray={currentShiftArray} />
    );
    expect(container).toMatchSnapshot();
    MockDate.reset();
  });
  it("should render shiftHours hours", () => {
    const currentShiftArray = [14, 15, 16, 17, 18, 19, 20, 21];
    const { getAllByTestId } = render(
      <CalendarHeader currentShiftArray={currentShiftArray} />
    );
    const hours = getAllByTestId("hour");
    expect(hours.length).toBe(currentShiftHoursArray().length);
  });
});
