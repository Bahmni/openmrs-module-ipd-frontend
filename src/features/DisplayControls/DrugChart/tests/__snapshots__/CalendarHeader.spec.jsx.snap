// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarHeader should match snapshot 1`] = `
<div>
  <div
    class="calendar-header"
  >
    <div
      style="display: flex;"
    >
      <div
        class="hour-header"
        data-testid="hour"
      >
        14:00
      </div>
      <div
        class="hour-header"
        data-testid="hour"
      >
        15:00
      </div>
      <div
        class="hour-header"
        data-testid="hour"
      >
        16:00
      </div>
      <div
        class="hour-header"
        data-testid="hour"
      >
        17:00
      </div>
      <div
        class="hour-header"
        data-testid="hour"
      >
        18:00
      </div>
      <div
        class="hour-header"
        data-testid="hour"
      >
        19:00
      </div>
      <div
        class="hour-header"
        data-testid="hour"
      >
        20:00
      </div>
      <div
        class="hour-header"
        data-testid="hour"
      >
        21:00
      </div>
    </div>
  </div>
</div>
`;
